trickle/trickle_server.go

```diff
@@ -0,0 +1,423 @@
+package trickle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+)
+
+// TODO sweep idle streams connections
+
+type StreamManager struct {
+	mutex   sync.RWMutex
+	streams map[string]*Stream
+}
+
+type Stream struct {
+	mutex       sync.RWMutex
+	segments    []*Segment
+	latestWrite int
+}
+
+type Segment struct {
+	idx    int
+	mutex  *sync.Mutex
+	cond   *sync.Cond
+	buffer *bytes.Buffer
+	closed bool
+}
+
+type SegmentSubscriber struct {
+	segment *Segment
+	readPos int
+}
+
+const maxSegmentsPerStream = 5
+
+const BaseServerPath = "/ai/live-video/"
+
+var FirstByteTimeout = errors.New("pending read timeout")
+
+func ConfigureServerWithMux(mux *http.ServeMux) {
+	/* TODO we probably want to configure the below
+	srv := &http.Server{
+		// say max segment size is 20 secs
+		// we can allow 2 * 20 secs given preconnects
+		ReadTimeout:  40 * time.Second,
+		WriteTimeout: 45 * time.Second,
+	}
+	*/
+
+	streamManager := &StreamManager{
+		streams: make(map[string]*Stream),
+	}
+	mux.HandleFunc("GET "+BaseServerPath+"{streamName}/{idx}", streamManager.handleGet)
+	mux.HandleFunc("POST "+BaseServerPath+"{streamName}/{idx}", streamManager.handlePost)
+	mux.HandleFunc("DELETE "+BaseServerPath+"{streamName}", streamManager.handleDelete)
+}
+
+func (sm *StreamManager) getStream(streamName string) (*Stream, bool) {
+	sm.mutex.RLock()
+	defer sm.mutex.RUnlock()
+	stream, exists := sm.streams[streamName]
+	return stream, exists
+}
+
+func (sm *StreamManager) getOrCreateStream(streamName string) *Stream {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	stream, exists := sm.streams[streamName]
+	if !exists {
+		stream = &Stream{
+			segments: make([]*Segment, 5),
+		}
+		sm.streams[streamName] = stream
+		slog.Info("Creating stream", "stream", streamName)
+	}
+	return stream
+}
+
+func (sm *StreamManager) clearAllStreams() {
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+
+	for _, stream := range sm.streams {
+		stream.clear()
+	}
+	sm.streams = make(map[string]*Stream)
+}
+
+func (s *Stream) clear() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	for _, segment := range s.segments {
+		segment.close()
+	}
+	s.segments = make([]*Segment, maxSegmentsPerStream)
+}
+
+func (sm *StreamManager) handleDelete(w http.ResponseWriter, r *http.Request) {
+	streamName := r.PathValue("streamName")
+	stream, exists := sm.getStream(streamName)
+	if !exists {
+		http.Error(w, "Invalid stream name", http.StatusBadRequest)
+		return
+	}
+
+	// TODO properly clear sessions once we have a good solution
+	//      for session reuse
+	return
+
+	stream.clear()
+	sm.mutex.Lock()
+	defer sm.mutex.Unlock()
+	delete(sm.streams, streamName)
+	slog.Info("Deleted stream", "streamName", streamName)
+}
+
+func (sm *StreamManager) handlePost(w http.ResponseWriter, r *http.Request) {
+	stream := sm.getOrCreateStream(r.PathValue("streamName"))
```

Comment:
IIUC here we'll need to handle payment (or at least check have some check if the segment is covered). Is it correct @j0sh ?

Reply:
Mostly yes, but not in this exact spot because not all streams will have payments, such as the  result stream.

I can add some middleware for the O to handle incoming segments and you can add in payments from there, eg by modifying outgoing headers. But if you are thinking about an out-of-band payments mechanism then the middleware can probably be skipped for now.

Reply:
Yeah, even if we do the payment outside of this flow, then we'll need to have somewhere the check "if it's paid" in order to stop sending segments if no payment.

Reply:
Yep - payments outside the flow can be a little simpler, eg a local (in-memory) subscriber rather than middleware that modifies the stream. Then you can gather any needed metrics within the subscribe callback and use those to check payments.

BTW if we do payments by "time" then it might not even be needed!